src/server/game/Accounts/AccountMgr.cpp

```diff
@@ -8,278 +8,451 @@
+    if (PreparedQueryResult result = LoginDatabase.Query(stmt))
+    {
+        Acore::Crypto::SRP6::Salt salt = (*result)[0].GetBinary<Acore::Crypto::SRP6::SALT_LENGTH>();
+        Acore::Crypto::SRP6::Verifier verifier = (*result)[1].GetBinary<Acore::Crypto::SRP6::VERIFIER_LENGTH>();
+        if (Acore::Crypto::SRP6::CheckLogin(username, password, salt, verifier))
+            return true;
+    }
+
+    return false;
+}
+
+uint32 AccountMgr::GetCharactersCount(uint32 accountId)
+{
+    // check character count
+    CharacterDatabasePreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_SEL_SUM_CHARS);
+    stmt->setUInt32(0, accountId);
+    PreparedQueryResult result = CharacterDatabase.Query(stmt);
+
+    return (result) ? (*result)[0].GetUInt64() : 0;
+}
+
+bool AccountMgr::IsPlayerAccount(uint32 gmlevel)
+{
+    return gmlevel == SEC_PLAYER;
+}
+
+bool AccountMgr::IsGMAccount(uint32 gmlevel)
+{
+    return gmlevel >= SEC_MODERATOR && gmlevel <= SEC_CONSOLE;
+}
+
+bool AccountMgr::IsAdminAccount(uint32 gmlevel)
+{
+    return gmlevel >= SEC_ADMINISTRATOR && gmlevel <= SEC_CONSOLE;
+}
+
+bool AccountMgr::IsConsoleAccount(uint32 gmlevel)
+{
+    return gmlevel == SEC_CONSOLE;
+}
+
+// RBAC
+void AccountMgr::LoadRBAC()
+{
+    ClearRBAC();
+
+    LOG_DEBUG("rbac", "AccountMgr::LoadRBAC");
+    uint32 oldMSTime = getMSTime();
+    uint32 count1 = 0;
+    uint32 count2 = 0;
+    uint32 count3 = 0;
+
+    LOG_DEBUG("rbac", "AccountMgr::LoadRBAC: Loading permissions");
+    QueryResult result = LoginDatabase.Query("SELECT id, name FROM rbac_permissions");
+    if (!result)
+    {
+        LOG_INFO("server.loading", ">> Loaded 0 account permission definitions. DB table `rbac_permissions` is empty.");
+        return;
     }
 
-    bool CheckPassword(uint32 accountId, std::string password)
+    do
+    {
+        Field* field = result->Fetch();
+        uint32 id = field[0].GetUInt32();
+        _permissions[id] = new rbac::RBACPermission(id, field[1].GetString());
+        ++count1;
+    } while (result->NextRow());
+
+    LOG_DEBUG("rbac", "AccountMgr::LoadRBAC: Loading linked permissions");
+    result = LoginDatabase.Query("SELECT id, linkedId FROM rbac_linked_permissions ORDER BY id ASC");
+    if (!result)
     {
-        std::string username;
+        LOG_INFO("server.loading", ">> Loaded 0 linked permissions. DB table `rbac_linked_permissions` is empty.");
+        return;
+    }
 
-        if (!GetName(accountId, username))
-            return false;
+    uint32 permissionId = 0;
+    rbac::RBACPermission* permission = nullptr;
 
-        Utf8ToUpperOnlyLatin(username);
-        Utf8ToUpperOnlyLatin(password);
+    do
+    {
+        Field* field = result->Fetch();
+        uint32 newId = field[0].GetUInt32();
+        if (permissionId != newId)
+        {
+            permissionId = newId;
+            permission = new rbac::RBACPermission(newId, field[1].GetString());
+            // permission = _permissions[newId];
```

Comment:
Okay dryrun fail because of this.

Original is permission = _permissions[newId]; but then worldserver crash on load.

I dno... :shrug:

Reply:
```cpp
do
    {
        Field* field = result->Fetch();
        uint32 newId = field[0].GetUInt32();
        if (permissionId != newId)
        {
            permissionId = newId;
            permission = _permissions[newId];
        }

        uint32 linkedPermissionId = field[1].GetUInt32();
        if (linkedPermissionId == permissionId)
        {
            LOG_ERROR("sql.sql", "RBAC Permission %u has itself as linked permission. Ignored", permissionId);
            continue;
        }
        permission->AddLinkedPermission(linkedPermissionId);
        ++count2;
    }
    while (result->NextRow());
```

Reply:
I can't do that

Reply:
with permission = _permissions[newId];
worldserver crashes on loading LoadRBAC()